recipes/openai.py

```diff
@@ -0,0 +1,378 @@
+    def from_prodigy(cls, example: Dict, labels: Iterable[str]) -> "PromptExample":
+        """Create a prompt example from Prodigy's format."""
+        pass
+
+
+def normalize_label(label: str) -> str:
+    return label.lower()
+
+
+class OpenAISuggester:
+    """Suggest annotations using OpenAI's ChatGPT
+
+    prompt_template (jinja2.Template): A Jinja2 template that contains the
+        prompt to send to OpenAI's ChatGPT API.
+    model (str): The ChatGPT model ID to use for completion. Check the OpenAI
+        documentation for more information https://beta.openai.com/docs/models/overview.
+    labels (List[str]): List of labels for annotation.
+    max_examples (int): The maximum number of examples to stream in the Prodigy UI.
+    segment (bool): If set to True, segment the documents into sentences.
+    verbose (bool): Show verbose output in the command-line, including the prompt and response from OpenAI.
+    openai_api_org (str): The OpenAI API organization.
+    openai_api_key (str): The OpenAI API key.
+    openai_temperature (float): The temperature parameter (from 0 to 1) that controls the
+        randomness of ChatGPT's output.
+    openai_max_tokens (int): The maximum amout of tokens that ChatGPT's
+        completion API can generate.
+    openai_n (int): The number of completions to generate for each prompt.
+    openai_n_retries (int): The number of retries whenever a 429 error occurs.
+    openai_retry_timeout_s (int): The amount of time before attempting another request whenever we
+        encounter a 429 error. Increases exponentially for each retry.
+    openai_read_timeout_s (int): The amount of time to wait a response output during a request.
+    examples (List[PromptExample]): A list of examples to add to the prompt to guide ChatGPT output.
+    response_parser (Callable[str] -> Dict): A function that accepts a string that represents
+        ChatGPT's raw response, and parses it into a dictionary that is compatible to Prodigy's
+        annotation interfaces.
+    """
+
+    prompt_template: jinja2.Template
+    model: str
+    labels: List[str]
+    max_examples: int
+    segment: bool
+    verbose: bool
+    openai_api_org: str
+    openai_api_key: str
+    openai_temperature: float
+    openai_max_tokens: int
+    openai_retry_timeout_s: int
+    openai_read_timeout_s: int
+    openai_n_retries: int
+    openai_n: int
+    examples: List[PromptExample]
+    response_parser: Callable
+
+    def __init__(
+        self,
+        prompt_template: jinja2.Template,
+        *,
+        labels: List[str],
+        max_examples: int,
+        segment: bool,
+        openai_api_org: str,
+        openai_api_key: str,
+        openai_model: str,
+        response_parser: Callable,
+        openai_temperature: int = 0,
+        openai_max_tokens: int = 500,
+        openai_retry_timeout_s: int = 1,
+        openai_read_timeout_s: int = 30,
+        openai_n_retries: int = 10,
+        openai_n: int = 1,
```

Comment:
I also made a few parameters more explicit. For example:
- `openai_retry_timeout_s`: is the amount of time to wait before trying the call again. This scales exponentially, i.e., 2x for every retry.
- `openai_read_timeout_s`: basically the amount of time to wait before passing a `httpx.ReadTimeOut` error. I increased this to 30 seconds.
- `openai_n_retries`: to avoid confusion, this parameter replaces `openai_n` (which dictates how many results to generate). This one controls how many times we will retry the call in the case of 429 errrors.
- `openai_n`: this one is more on ChatGPT parameters. Before we were using this for the number of retries but it might help to be more explicit.